### PR TITLE
ENG-1427 - Add sorting by seven-day VP change

### PR DESCRIPTION
**Pull request by jefag**
This PR adds a competing implementation for #58 , the main-difference is this approach uses block-headers for time tracking, so there is no scheduler, no "re-calc" logic needed, no complexity of dependency injection (passing the client to the Data Product).

It's late, and I'm in a bit of a rush, so I need to check for off-by-one bugs, edge cases, and add tests still...but the algo shows promise.

# Performance 

- zero-latency
- 320 ms round-trip (roughly ~250 ms server side) load for sorting on VP change on first pass, then 160 ms (roughly 100 ms server side) for any there-after in the same hour, without the actual delta cached (only cacheing the 7-day VP for 1-hour). So, these numbers will be even better in the cluster.

### CODE REVIEW

File: app/data_products.py [hunk 1]
```diff
@@ -240,7 +273,9 @@ def handle(self, event):
                     
                     # Update voting power
                     self.delegatee_vp[old_delegate] -= amount
-                    self.delegatee_vp_history[old_delegate].append((block_number, self.delegatee_vp[old_delegate]))
+
+                    # TODO: Is this redundant? With the line below during handling of 'DELEGATE_VOTES_CHANGE'?
+                    # self.delegatee_vp_history[old_delegate].append((block_number, self.delegatee_vp[old_delegate]))
```
Comment (jefag) [inline app/data_products.py:278]: @CarlGocht I could use a sanity check, that we don't need lines 278 and and 299.

File: tests/test_endpoints.py [hunk 2]
```diff
@@ -108,16 +108,16 @@ async def test_delegates_endpoint_sort_by_oldest(app):
...
     assert len(delegates) == 3
     assert delegates[0]["addr"] == "0x2222222222222222222222222222222222222222"  # Block 200 (highest)
-    assert delegates[0]["oldest_block"] == 200
+    assert delegates[0]["OLD"] == 200
     assert delegates[1]["addr"] == "0x1111111111111111111111111111111111111111"  # Block 100
-    assert delegates[1]["oldest_block"] == 100
+    assert delegates[1]["OLD"] == 100
     assert delegates[2]["addr"] == "0x3333333333333333333333333333333333333333"  # Block 50 (lowest)
-    assert delegates[2]["oldest_block"] == 50
+    assert delegates[2]["OLD"] == 50
```
Comment (jefag) [inline tests/test_endpoints.py:115]: @CarlGocht One change is that the endpoint returns fields which match the request.

Hopefully this is trivial to change in the agora-next client.

File: tests/test_endpoints.py [hunk 3]
```diff
@@ -243,46 +243,44 @@ def __init__(self):
...
-    assert delegates[0][0] == '0x5555'
-    assert delegates[0][1] == 500
-    assert delegates[1][0] == '0x4444'
-    assert delegates[1][1] == 400
-    assert delegates[2][0] == '0x2222'
-    assert delegates[2][1] == 200
-    assert delegates[3][0] == '0x1111'
-    assert delegates[3][1] == 100
-    assert delegates[4][0] == '0x3333'
-    assert delegates[4][1] == 0
```
Comment (jefag) [inline tests/test_endpoints.py:256]: @CarlGocht , I just noticed moments ago, that the tests were using positions for the fields rather than the field names (ie `delegates[0][0]` rather than `delegates[0]['addr']`).  

Was the tip of main serving data by position or field, in the actual endpoint?  

I thought it was always using fields, but it would actually be faster and lower data to transmit to give the results by position.  Anyway flagging this, as it might be a source of confusion for you or I.

**Comment by jefag** [comment]
@CarlGocht From [524aebd](https://github.com/voteagora/dao-node/pull/72/commits/524aebd51e3c49d8c0c3d1f1a56949b1a20c6b9a) it should boot for optimism and ENS.

**Comment by jefag** [comment]
...but I did run out of time.  Could easy be a gotcha somewhere I haven't caught yet.

File: app/data_products.py (hunk 1 above)
Comment (CarlGocht) [reply app/data_products.py:278]: yes. we do not need these lines if we have the new version of 7d change you have here

File: tests/test_endpoints.py (hunk 3 above)
Comment (CarlGocht) [reply tests/test_endpoints.py:256]: ya it was using fields. will keep an eye out for any issues with that

**Review by pedrohasf — APPROVED** [review]

File: app/server.py [hunk 4]
```diff
@@ -728,198 +725,59 @@ async def delegates_handler(app, request):
...
-            out = [{'addr': obj[0], 'last_vote_block': obj[1],
-                   'most_recent_block': app.ctx.delegations.delegatee_latest_event[obj[0]]['block_number'] if obj[0] in app.ctx.delegations.delegatee_latest_event else None} for obj in out]
+    # Everything below, replaces that gawdaful combinatorical mess with a single dict(**(k, func(v)),
+    # where func(v) is a lambda that returns the value for key k, but ... this might actually be slower
+    # than writing a million if-statements out, per-row, because we're iterating over transformers
+    # dynamically at runtime, rather than at compile time.  Would love to profile the diff...
+    # but alas, I'm out of time...
+
+    # If this is terribly slow...backup plan is now the idea Carl had, and if both of these
+    # are too slow, then Jeff will write the million if-statements out.
```
Comment (Daktic) [inline app/server.py:750]: What are the backup plans? Could also look into Rython or codegen

Comment (jefag) [reply app/server.py:750]: Turns out, it's not terribly slow!

Comment (jefag) [reply app/server.py:750]: w00t! I'll remove this comment.

File: app/data_products.py [hunk 5]
```diff
@@ -155,10 +174,31 @@ def _parse_delegate_array(self, array_str):
...
+
+        assert isinstance(block_number, int)
+        assert isinstance(timestamp, int)
+
+        self.current_block_number = block_number
+        self.current_ts = timestamp
+
+        self.timestamp_to_block[timestamp] = block_number
 
+        rounded_ts = timestamp - (timestamp % 3600)
```
[the remark this answers is not in the document]
Comment (jefag) [reply app/data_products.py:188]: I think it's just `round`, but point taken.  I'll fix.

[the remark this answers is not in the document]
Comment (jefag) [reply app/data_products.py:188]: https://github.com/voteagora/dao-node/pull/74/commits/d60e8aa204b643712191ddf5b1182261ad72c4a7

File: app/server.py (hunk 4 above)
Comment (jefag) [reply app/server.py:750]: https://github.com/voteagora/dao-node/pull/74/commits/d60e8aa204b643712191ddf5b1182261ad72c4a7
